File: 160.相交链表/getIntersectionNode.go
package problem0160

type ListNode struct {
	Val  int
	Next *ListNode
}

// 相交链表：有同样的末尾, 相同的末尾是关键
// headA: ----@@@@
// headB: --@@@@
// headA+headB: ----@@@@--@@@@
// headB+headA: --@@@@----@@@@
func getIntersectionNode(headA, headB *ListNode) *ListNode {
	if headA == nil || headB == nil {
		return nil
	}
	tA, tB := headA, headB
	for tA != tB {
		if tA != nil {
			tA = tA.Next
		} else {
			tA = headB
		}
		if tB != nil {
			tB = tB.Next
		} else {
			tB = headA
		}
	}
	return tA
}

File: 160.相交链表/getIntersectionNode.js
// 相交链表
// 找到两个单链表相交的起始节点
// 如果两个链表没有交点，返回 null

// 相交链表 = 有公共末尾
// headA = ____@@@@
// headB = __@@@@
// headA+headB = ____@@@@__@@@@
// headB+headA = __@@@@____@@@@
// 寻找相同后缀
function getIntersectionNode(headA, headB) {
	if (!headA || !headB) {
		return null
	}
	let a = headA
	let b = headB
	while (a !== b) {
		if (a !== null) {
			a = a.next
		} else {
			a = headB
		}
		if (b !== null) {
			b = b.next
		} else {
			b = headA
		}
	}
	return a
}